File: dataset_construction/flow_extraction.rb
require 'csv'
require 'optparse'

# Flow gaps in microseconds
FLOW_GAPS = [500, 5000, 10000]
DEFAULT_IP = 192.168.60.81

# Parse "HH:MM:SS.ssssss" -> float seconds
def parse_time(ts)
  h, m, s = ts.split(":")
  h.to_i * 3600 + m.to_i * 60 + s.to_f
end

# Binary-search helper: given a sorted timeline [[t0, v0], [t1, v1], ...],
# return cumulative value at time <= t (or 0 if none).
def cumulative_at(timeline, t)
  return 0 if timeline.nil? || timeline.empty?
  lo, hi = 0, timeline.length - 1
  ans = -1
  while lo <= hi
    mid = (lo + hi) / 2
    if timeline[mid][0] <= t
      ans = mid
      lo = mid + 1
    else
      hi = mid - 1
    end
  end
  return 0 if ans < 0
  timeline[ans][1]
end

def extract_flows(input_file, output_file, flow_gap_us, vantage_ip)
  flow_gap_s = flow_gap_us.to_f / 1_000_000.0
  global_flow_id = 0

  groups = Hash.new { |h, k| h[k] = [] }
  events = []

  CSV.foreach(input_file, headers: true) do |row|
    t = parse_time(row["timestamp"])
    src_ip = row["source_ip"]
    dst_ip = row["destination_ip"]
    size = row["packet_size"].to_i

    key = [src_ip, row["source_port"], dst_ip, row["destination_port"]]
    groups[key] << { time: t, size: size, ts_raw: row["timestamp"] }

    events << { t: t, src: src_ip, dst: dst_ip, size: size }
  end

  # Build cumulative incoming/outgoing timelines per IP
  events.sort_by! { |e| e[:t] }
  outgoing_tl = Hash.new { |h, k| h[k] = [] }
  incoming_tl = Hash.new { |h, k| h[k] = [] }
  out_cum = Hash.new(0)
  in_cum = Hash.new(0)

  events.each do |e|
    out_cum[e[:src]] += e[:size]
    in_cum[e[:dst]] += e[:size]
    outgoing_tl[e[:src]] << [e[:t], out_cum[e[:src]]]
    incoming_tl[e[:dst]] << [e[:t], in_cum[e[:dst]]]
  end

  CSV.open(output_file, "w") do |csv|
    csv << [
      "flow_id", "src_ip", "src_port", "dst_ip", "dst_port",
      "start_time", "end_time", "packet_count", "total_size",
      "gap_us", "neworkin", "neworkout"
    ]

    groups.each do |(src_ip, src_port, dst_ip, dst_port), packets|
      packets.sort_by! { |p| p[:time] }

      current_flow = []
      last_pkt_time = nil
      prev_flow_end = nil

      packets.each do |pkt|
        if last_pkt_time && (pkt[:time] - last_pkt_time > flow_gap_s)
          start_t  = current_flow.first[:time]
          end_t    = current_flow.last[:time]
          start_ts = current_flow.first[:ts_raw]
          end_ts   = current_flow.last[:ts_raw]
          gap_us   = prev_flow_end.nil? ? 0 : ((start_t - prev_flow_end) * 1_000_000).round

          # Determine the reference IP
          ref_ip = vantage_ip || src_ip
          net_in  = cumulative_at(incoming_tl[ref_ip], start_t - 1e-12)
          net_out = cumulative_at(outgoing_tl[ref_ip], start_t - 1e-12)

          csv << [
            global_flow_id, src_ip, src_port, dst_ip, dst_port,
            start_ts, end_ts, current_flow.size, current_flow.sum { |x| x[:size] },
            gap_us, net_in, net_out
          ]

          global_flow_id += 1
          prev_flow_end = end_t
          current_flow = []
        end

        current_flow << pkt
        last_pkt_time = pkt[:time]
      end

      unless current_flow.empty?
        start_t  = current_flow.first[:time]
        end_t    = current_flow.last[:time]
        start_ts = current_flow.first[:ts_raw]
        end_ts   = current_flow.last[:ts_raw]
        gap_us   = prev_flow_end.nil? ? 0 : ((start_t - prev_flow_end) * 1_000_000).round

        ref_ip = vantage_ip || src_ip
        net_in  = cumulative_at(incoming_tl[ref_ip], start_t - 1e-12)
        net_out = cumulative_at(outgoing_tl[ref_ip], start_t - 1e-12)

        csv << [
          global_flow_id, src_ip, src_port, dst_ip, dst_port,
          start_ts, end_ts, current_flow.size, current_flow.sum { |x| x[:size] },
          gap_us, net_in, net_out
        ]
        global_flow_id += 1
      end
    end
  end
end

if __FILE__ == $0
  script_dir = File.expand_path(File.dirname(__FILE__))
  input_file = File.join(script_dir, "../trace_collector/packet_traces/packets.csv")

  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: ruby flow_extractor.rb [options]"
    opts.on("--ip IP", "Specific vantage IP address") { |v| options[:ip] = v }
  end.parse!

  vantage_ip = options[:ip] || DEFAULT_IP
  puts "Using vantage IP: #{vantage_ip || '(per-flow source_ip)'}"

  FLOW_GAPS.each do |gap|
    output_file = File.join(script_dir, "flows_#{gap}.csv")
    extract_flows(input_file, output_file, gap, vantage_ip)
    puts "Generated #{output_file}"
  end
end

File: dataset_construction/flow_extraction.py
#!/usr/bin/env python3
import csv, os
from bisect import bisect_right

FLOW_GAP_US = 5000
VANTAGE_IP = "192.168.60.81"

def cumulative_at(timeline, t_ns):
    if not timeline: return 0
    times = [x[0] for x in timeline]
    idx = bisect_right(times, t_ns) - 1
    return 0 if idx < 0 else timeline[idx][1]

def extract_flows(input_file, output_file, flow_gap_us, vantage_ip):
    flow_gap_ns = flow_gap_us * 1000
    fid = 0
    groups = {}
    events = []

    with open(input_file, "r", newline="", encoding="utf-8") as f:
        r = csv.DictReader(f)
        for row in r:
            t_ns = int(row["ts_ns"])
            ts_raw = row.get("timestamp_epoch_s","")
            src_ip, dst_ip = row["source_ip"], row["destination_ip"]
            key = (src_ip, row["source_port"], dst_ip, row["destination_port"])
            pkt = {"t_ns": t_ns, "size": int(row["packet_size"]), "ts_raw": ts_raw}
            groups.setdefault(key, []).append(pkt)
            events.append({"t_ns": t_ns, "src": src_ip, "dst": dst_ip, "size": pkt["size"]})

    events.sort(key=lambda e: e["t_ns"])
    outgoing_tl, incoming_tl, out_cum, in_cum = {}, {}, {}, {}
    for e in events:
        out_cum[e["src"]] = out_cum.get(e["src"], 0) + e["size"]
        in_cum[e["dst"]] = in_cum.get(e["dst"], 0) + e["size"]
        outgoing_tl.setdefault(e["src"], []).append((e["t_ns"], out_cum[e["src"]]))
        incoming_tl.setdefault(e["dst"], []).append((e["t_ns"], in_cum[e["dst"]]))

    with open(output_file, "w", newline="", encoding="utf-8") as fo:
        w = csv.writer(fo)
        w.writerow([
            "flow_id","src_ip","src_port","dst_ip","dst_port",
            "start_time","end_time","start_ts_ns",
            "packet_count","total_size","gap_us","neworkin","neworkout"
        ])

        for (src_ip, src_port, dst_ip, dst_port), packets in groups.items():
            packets.sort(key=lambda p: p["t_ns"])
            cur = []
            last_ns = prev_end_ns = None

            def flush(cur, prev_end_ns):
                nonlocal fid
                if not cur: return prev_end_ns
                start_ns, end_ns = cur[0]["t_ns"], cur[-1]["t_ns"]
                start_ts, end_ts = cur[0]["ts_raw"], cur[-1]["ts_raw"]
                gap_us = 0 if prev_end_ns is None else (start_ns - prev_end_ns) // 1000
                ref_ip = vantage_ip or src_ip
                net_in = cumulative_at(incoming_tl.get(ref_ip), start_ns - 1)
                net_out = cumulative_at(outgoing_tl.get(ref_ip), start_ns - 1)
                w.writerow([
                    fid, src_ip, src_port, dst_ip, dst_port,
                    start_ts, end_ts, start_ns,
                    len(cur), sum(x["size"] for x in cur),
                    gap_us, net_in, net_out
                ])
                fid += 1
                return end_ns

            for pkt in packets:
                if last_ns is not None and pkt["t_ns"] - last_ns > flow_gap_ns:
                    prev_end_ns = flush(cur, prev_end_ns)
                    cur = []
                cur.append(pkt)
                last_ns = pkt["t_ns"]

            prev_end_ns = flush(cur, prev_end_ns)

def main():
    base = os.path.dirname(os.path.abspath(__file__))
    input_file = os.path.join(base, "../trace_collector/packet_traces.csv")
    output_file = os.path.join(base, "flow_extraction.csv")
    extract_flows(input_file, output_file, FLOW_GAP_US, VANTAGE_IP)

if __name__ == "__main__":
    main()
